mlir/lib/Transforms/Utils/DialectConversion.cpp

```diff
@@ -2865,6 +2860,8 @@ static LogicalResult legalizeUnresolvedMaterialization(
       break;
     }
     if (newMaterialization) {
+      assert(newMaterialization.getType() == opResult.getType() &&
+             "materialization callback produced value of incorrect type");
```

Comment:
Assuming this assert might be hit more often now, it might be worth duplicating it into the respective switch cases so that we can give a better error message such as 
`"value returned by argument materialization expected to in the source type system"` that are more actionable

Reply:
Another alternative would be to insert the cast here, if the types don't match. That'll probably be easier to debug in the IR when it goes wrong, but not sure if it fits in the design of the API.

Reply:
`source type system` is not really the correct word. In case of an argument materialization, we have to produce an SSA value of the original type. That original type could an illegal type, but it could also be a legal type. (It is triggered by the user through `applySignatureConversion` and it is possible to convert a signature that's already legal.)

I clarified the comment in the type converter a bit.

> Another alternative would be to insert the cast here, if the types don't match.

That would work indeed but we also don't insert `unrealized_conversion_cast` during `RewriterBase::replaceOp` when the types of the replacement values do not match.